editoast/src/views/projection.rs

```diff
@@ -290,96 +269,108 @@ pub async fn compute_projected_train_paths(
-        &trains_schedules,
+        &train_schedules,
         infra,
         electrical_profile_set_id,
     )
     .await?;
 
     // 2. Extracts train simulation details and computes unique hashes for projected train paths.
-    let trains_details = extract_train_details(&trains_schedules, simulations).await?;
+    let trains_details = extract_train_details(simulations).await?;
 
-    let mut trains_hash_values = vec![];
+    let mut trains_hash_values: HashMap<String, Vec<usize>> = HashMap::new();
 
-    trains_hash_values.extend(trains_details.iter().map(|t| {
-        t.compute_projection_hash_with_versioning(infra.id, infra.version, &track_section_ranges)
-    }));
+    trains_details
+        .iter()
+        .enumerate()
+        .for_each(|(index, train_details)| {
+            train_details.as_ref().map(|train_details| {
+                let hash = train_details.compute_projection_hash_with_versioning(
+                    infra.id,
+                    infra.version,
+                    &track_section_ranges,
+                );
+                trains_hash_values.entry(hash).or_default().push(index);
+            });
+        });
+
+    let trains_hash_values_to_vec: Vec<_> = trains_hash_values.keys().cloned().collect();
 
     // 3. Retrieve cached projection
-    let (miss_cache, mut hit_cache) =
-        retrieve_cached_projection(&mut valkey_conn, &trains_hash_values, &trains_details).await?;
 
-    // 4. Compute space time curves for all miss cache
-    let space_time_curves = compute_batch_space_time_curves(&miss_cache, &path_projection).await;
+    let cached_projections = valkey_conn
+        .json_get_bulk(&trains_hash_values_to_vec)
+        .await?
+        .collect::<Vec<Option<SpaceTimeCurves>>>();
 
-    // 5. Store the projection in the cache (using pipeline)
-    let trains_hash_values: HashMap<_, _> = trains_details
-        .iter()
-        .map(|t| t.train_id)
-        .zip(trains_hash_values)
-        .collect();
-    let mut new_items = vec![];
-    for train_id in miss_cache.iter().map(|t| t.train_id) {
-        let hash = &trains_hash_values[&train_id];
-        let cached_value = space_time_curves
-            .get(&train_id)
-            .expect("Space time curves not available for train")
-            .clone();
-        hit_cache.push((cached_value.clone(), train_id));
-        new_items.push((hash, cached_value));
+    let mut projection_request_map: HashMap<String, TrainSimulationDetails> = HashMap::new();
+    let mut project_path_result: Vec<Arc<SpaceTimeCurves>> =
+        vec![Arc::default(); train_schedules.len()];
+    for (hash, projection) in trains_hash_values_to_vec
+        .into_iter()
+        .zip(cached_projections)
+    {
+        if let Some(projection) = projection {
+            let indexes = &trains_hash_values[&hash];
+            let projection = Arc::new(projection.clone());
+            for index in indexes {
+                project_path_result[*index] = projection.clone();
```

Comment:
You should be able to remove those 2 `clone`. The first one, you give ownership of the cached projection to the `Arc` and don't use the cached projection anymore. The second one, you give the `Arc` to `project_path_result`  and don't use after.

```suggestion
            let projection = Arc::new(projection);
            for index in indexes {
                project_path_result[*index] = projection;
```

Reply:
For the second one, we're in a loop, so it requires cloning it.

Reply:
I agree for the first one but the second one is inside a loop !

Reply:
oops !